File: backend/uploads/LargerThanThreshold.m
function result = LargerThanThreshold(channel_key, threshold)
    % Read JSON data from file based on channel_key
    % Assuming channel_key is the filename
    % 动态生成 URL 和查询参数
    % url = sprintf('https://localhost:5000/api/channel-data/?channel_key=%s', channel_key);
    
    % 设置请求选项
    % options = weboptions('ContentType', 'json');
    
    % 发送 GET 请求
    %json_data = webread(url, options);
    %fprintf('运行时间：%.4f 秒\n', elapsedTime);
    json_data = channel_key

    % Extract X_value and Y_value from the JSON data
    X_value = json_data.X_value;
    Y_value = json_data.Y_value;

    % Find the ranges of X_value where Y_value is greater than the threshold
    above_threshold = abs(Y_value) > threshold;
    ranges = [];
    start_idx = -1;

    for i = 1:length(above_threshold)
        if above_threshold(i) && start_idx == -1
            % Start of a new range
            start_idx = i;
        elseif ~above_threshold(i) && start_idx ~= -1
            % End of the current range
            ranges = [ranges; X_value(start_idx), X_value(i-1)];
            start_idx = -1;
        end
    end

    % If the last value was above threshold, close the range
    if start_idx ~= -1
        ranges = [ranges; X_value(start_idx), X_value(end)];
    end

    % Prepare JSON output
    result_struct.X_range = ranges;
    result = jsonencode(result_struct);
end

File: backend/uploads/LargerThanThreshold.py
import numpy as np

def LargerThanThreshold(channel_key, threshold):
    """
    Find ranges of X_value where Y_value is greater than the threshold.
    
    Parameters:
    channel_key (dict): Dictionary containing 'X_value' and 'Y_value' arrays
    threshold (float): Threshold value to compare against
    
    Returns:
    str: JSON string containing the X_value ranges where Y_value > threshold
    """
    # Extract X_value and Y_value from the input data
    X_value = np.array(channel_key['X_value'])
    Y_value = np.array(channel_key['Y_value'])
    
    # Find where Y_value is above threshold
    above_threshold = np.abs(Y_value) > threshold
    ranges = []
    start_idx = -1
    
    for i in range(len(above_threshold)):
        if above_threshold[i] and start_idx == -1:
            # Start of a new range
            start_idx = i
        elif not above_threshold[i] and start_idx != -1:
            # End of the current range
            ranges.append([X_value[start_idx], X_value[i-1]])
            start_idx = -1
    
    # If the last value was above threshold, close the range
    if start_idx != -1:
        ranges.append([X_value[start_idx], X_value[-1]])
    
    # Prepare output
    result_struct = {'X_range': ranges}
    return result_struct